### ISSUE TITLE
Proposal for a better to read and write case expression

### ISSUE THREAD

**Issue by burzum** [cakephp/cakephp#11793]
When I read this https://stackoverflow.com/questions/49129058/cakephp-3-query-ifnull I remembered that I disliked this, it's aweful to read and remember.

```php
$query
    ->newExpr()
    ->addCase(
        [
            $query->newExpr()->add([
                'Sales.isforeign' => $isforeign,
                'Sales.source' => $source
            ])
        ],
        [1, 0],
        ['integer', 'integer']
    );
```

So I had a discussion with @lorenzo about this and we agreed to implement something like this:

```php
$case = $query->newCase()
    ->when([
        'Sales.isforeign' => $isforeign,
        'Sales.source' => $source
    ])
    ->then(1)
    ->otherWise('0', 'integer');

$query
    ->find()
    ->addCase($case);
```

The then() and otherWise() *can* take a 2nd arg to define the type. If *not* set the type of the first arg is used by default.

Any other ideas or improvements please?

**Comment by josegonzalez** [comment]
Can you redo that case api with the example you have in the first bit?

**Comment by burzum** [comment]
@josegonzalez done.

**Comment by inoas** [comment]
shouldn't it be `->then(1, 'integer')` then?

**Comment by burzum** [comment]
@ionas 

>  The then() and otherWise() can take a 2nd arg to define the type. If not set the type of the first arg is used by default.

**Comment by raul338** [comment]
I would prefer define the condition and the return in the same call, like this
```php
$query->select([
    'late_status' => $query->newCase()
        ->add(
            $query->newExpr()->lt('task.diffStart', 0),
            'Sooner',
            'literal'
        )
        ->add(
            $query->newExpr()->between('task.diffStart', 0, 15),
            'In Time',
            'literal'
        )
        ->add(
            $query->newExpr()->gt('task.diffStart', 15),
            'Late'  // literal optional
        )
        ->else(
            'Task not started / Canceled',
            'literal'
        )
]);
```

**Comment by markstory** [comment]
@burzum What would multiple cases look like with that API? Should `CaseExpression` be exposed via a builder on `QueryExpression`? The `newCase()` method proposed by @raul338 would be a nice way to not require users to `new` an expression object.

**Comment by lorenzo** [comment]
For multiple ases I would have a completely different expression class. The syntax is so different that we could say they are different things

**Comment by markstory** [comment]
Having 2 expression classes would make deprecating the old one easier.

**Comment by dereuromark** [comment]
The PR is open but not further finished. We should either close the PR or this ticket and finish PR for merging.

**Comment by raul338** [comment]
I closed the PR as I don't feel there was a consensus on the new API for case statements.
Also I refactored some queries and I realised all I needed was an IF sql-statement that can be done with current api and doesn't look so bad.

```php
$query->newExpr()->addCase([$conditionExpr], [$trueExpr, $falseExpr]);
```

I don't know if I should open a PR to add that behavior.
I still believe the syntax should be like my previous comment.
